### PR TITLE
core: change return type of irq_is_in(), irq_is_enabled() to bool

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

This only cleans up the API, no change in behavior or users of the API is expected.

### Testing procedure

This is mostly a cosmetic change, if everything still compiles we are good.

### CODE REVIEW

**Comment by gschorcht** [comment]
Looks good. Do we need a second ACK because of API change?

**Review by gschorcht — APPROVED** [review]

**Comment by kaspar030** [comment]
does the code actually change with this? I always had the (tingling) fear that the compiler does some needless masking when the type is not "int".

**Comment by benpicco** [comment]
I get 

#### master

#### samr21-xpro
```
   text	   data	    bss	    dec	    hex	filename
  99808	    188	  18968	 118964	  1d0b4	/home/benpicco/dev/RIOT-BHT/examples/gnrc_networking/bin/samr21-xpro/gnrc_networking.elf
make BOARD=samr21-xpro RIOT_VERSION=test -j  37,56s user 3,49s system 648% cpu 6,329 total
```
#### avr-rss2
```
   text	   data	    bss	    dec	    hex	filename
 116284	  13642	  13467	 143393	  23021	/home/benpicco/dev/RIOT-BHT/examples/gnrc_networking/bin/avr-rss2/gnrc_networking.elf
make BOARD=avr-rss2 RIOT_VERSION=test -j  13,59s user 1,88s system 639% cpu 2,419 total
```

#### this PR

#### samr21-xpro
```
   text	   data	    bss	    dec	    hex	filename
  99808	    188	  18968	 118964	  1d0b4	/home/benpicco/dev/RIOT-BHT/examples/gnrc_networking/bin/samr21-xpro/gnrc_networking.elf
make BOARD=samr21-xpro RIOT_VERSION=test -j  34,50s user 3,03s system 656% cpu 5,713 total
```

#### avr-rss2
```
   text	   data	    bss	    dec	    hex	filename
 116284	  13642	  13467	 143393	  23021	/home/benpicco/dev/RIOT-BHT/examples/gnrc_networking/bin/avr-rss2/gnrc_networking.elf
make BOARD=avr-rss2 RIOT_VERSION=test -j  10,64s user 1,28s system 632% cpu 1,883 total
```

File: cpu/cortexm_common/include/irq_arch.h [hunk 1]
```diff
@@ -72,7 +73,7 @@ static inline __attribute__((always_inline)) int irq_is_enabled(void)
 /**
  * @brief See if the current context is inside an ISR
  */
-static inline __attribute__((always_inline)) int irq_is_in(void)
+static inline __attribute__((always_inline)) bool irq_is_in(void)
 {
     return (__get_IPSR() & 0xFF);
```
Comment (kaspar030) [inline cpu/cortexm_common/include/irq_arch.h:78]: this loses information if mapped to bool?

File: cpu/mips32r2_common/irq_arch.c [hunk 2]
```diff
@@ -35,11 +35,12 @@ void irq_restore(unsigned int state)
...
+bool irq_is_in(void)
 {
-    return (mips32_get_c0(C0_STATUS) & SR_EXL) != 0;
+    return mips32_get_c0(C0_STATUS) & SR_EXL;
 }
 
-int irq_is_enabled(void) {
+bool irq_is_enabled(void)
+{
     return mips32_get_c0(C0_STATUS) & SR_IE;
```
Comment (kaspar030) [inline cpu/mips32r2_common/irq_arch.c:45]: and this?

File: cpu/cortexm_common/include/irq_arch.h (hunk 1 above)
Comment (benpicco) [reply cpu/cortexm_common/include/irq_arch.h:78]: It only ever was intended to return non-0 for `true` and 0 for `false`. 
There is nothing that does (or should) rely on the implementation specific value this returns.

File: cpu/msp430_common/include/irq_arch.h [hunk 3]
```diff
@@ -83,12 +84,12 @@ __attribute__((always_inline)) static inline void irq_restore(unsigned int state
...
+__attribute__((always_inline)) static inline bool irq_is_in(void)
 {
     return __irq_is_in;
 }
 
-__attribute__((always_inline)) static inline int irq_is_enabled(void)
+__attribute__((always_inline)) static inline bool irq_is_enabled(void)
 {
     unsigned int state;
     __asm__ volatile(
```
Comment (kaspar030) [inline cpu/msp430_common/include/irq_arch.h:95]: `vendor/msp430-gcc-support-files/include/msp430f1611.h:#define GIE                    (0x0008)`

so, `state & GIE` is either 0 or 0x8 ...

Comment (benpicco) [reply cpu/msp430_common/include/irq_arch.h:95]: so?

File: cpu/cortexm_common/include/irq_arch.h (hunk 1 above)
Comment (kaspar030) [reply cpu/cortexm_common/include/irq_arch.h:78]:
```
   71 /**                                                                                                                                          
   72  * @brief   Test if IRQs are currently enabled                                                                                               
   73  *                                                                                                                                           
   74  * @warning Use this function from thread context only. When used in interrupt                                                               
   75  *          context the returned state may be incorrect.                                                                                     
   76  *                                                                                                                                           
   77  * @return  0 (false) if IRQs are currently disabled                                                                                         
   78  * @return  != 0 (true) if IRQs are currently enabled                                                                                        
   79  */                                                                                                                                          
   80 MAYBE_INLINE int irq_is_enabled(void);                                                                                                       
```

It specifically says `!= 0` in the docs. :/

**Comment by kaspar030** [comment]
This won't work. this is not bool for a reason (no need for extra cycles to map some register bit to 0 or 1).

File: cpu/cortexm_common/include/irq_arch.h (hunk 1 above)
Comment (benpicco) [reply cpu/cortexm_common/include/irq_arch.h:78]: `true` is `!= 0`
It does not specify which `!= 0` value to expect

**Comment by kaspar030** [comment]
> huh? Nothing says that `true` must be 1.
> `true` can be any value that is not `0`

I think you are mistaken:

(from https://pubs.opengroup.org/onlinepubs/007904975/basedefs/stdbool.h.html)
```
The <stdbool.h> header shall define the following macros:

bool
    Expands to _Bool.
true
    Expands to the integer constant 1.
false
    Expands to the integer constant 0.
```

so, technically, "bool" is some int (char? int? impementation defined?), but "true" and "false" are defined to "1" and "0".
and, the C compiler will happily evaluate every bool-int-type != 0 to logical true, but not `(any_int != 0) == true (literal)`. There is no "true" from the compiler perspective, it is defined to be "1" in some headers.

IMO, while this may seem as a cleanup, this is opening a can of worms.

File: cpu/msp430_common/include/irq_arch.h (hunk 3 above)
Comment (maribu) [reply cpu/msp430_common/include/irq_arch.h:95]: The compiler will convert any non-zero value to `1` if the type is `bool`. Note that with static inline functions this conversion can be optimised out if it is not needed (e.g. a branch if non-zero is used on the return value).

I tested a few years ago of using `bool` over `int` adds overhead for `static inline` functions, and for my test codes this wasn't the case.

**Comment by benpicco** [comment]
But this assumes the compiler has no idea about C99 `bool` types and it's all just preprocessor magic.

Code like this

```
static inline bool _true_func(void)
{
    return 123;
}

int main(int argc, char** argv)
{
    assert(1 == true);
    assert(_true_func() == true);
}
```

will work as one would expect. Only `assert(true == 123)` will not work - but that's because `true` gets promoted to an integer.

Also see https://godbolt.org/z/dEMYEExjE

**Comment by kaspar030** [comment]
Ok, you're right. This is the relevant [C standard quote](http://port70.net/~nsz/c/c11/n1570.html#6.3.1.2p1):

```
6.3.1.2 Boolean type

1 When any scalar value is converted to _Bool, the result is 0 if the value compares equal to 0; otherwise, the result is 1.59) 
```

File: cpu/cortexm_common/include/irq_arch.h (hunk 1 above)
Comment (gschorcht) [reply cpu/cortexm_common/include/irq_arch.h:78]: According to the C standard, `true` and `false` are macros that are 1 and 0. `bool` is also a macro and is simply an alias for `_Bool`. Furthermore, anything that is not 0 becomes 1 when assigned to `_Bool`. That is, whatever integer value is returned will become 1 if it is not 0. So if you assign any integer values that are not 0 to two boolean variables, the comparison will always be correct.

Maybe, the documentation should be changed to clarify that
```
 * @return  false (0) if IRQs are currently disabled                                                                                         
 * @return  true (1) if IRQs are currently enabled
```

**Comment by gschorcht** [comment]
> ```
> int main(int argc, char** argv)
> {
>     assert(1 == true);
>     assert(_true_func() == true);
> }
> ```
> 
> will work as one would expect. Only `assert(true == 123)` will not work - but that's because `true` gets promoted to an integer.

Exactly. With the former API, the only use is something like
```
if (irq_is_in()) {
```
We shouldn't suppose that someone used something like
```
if (irq_is_in() == SOME_IMPLEMENTATION_SPECIFIC) {
```
especially because our documentation of [`irq_is_in`](https://doc.riot-os.org/group__core__irq.html#ga83decbeef665d955290f730125ef0e3f) says that this function returns `false` or `true`. If we interpret this very strictly, someone could have tested `true` or `1` at most.

I'm still for changing it :+1:

File: cpu/cortexm_common/include/irq_arch.h (hunk 1 above)
Comment (maribu) [reply cpu/cortexm_common/include/irq_arch.h:78]: 👍 but better use `@retval` for documenting individual return values.

**Review by aabadie — APPROVED** [review]
Seems like there is a consensus on this change.

Here is a second ACK :)

**Comment by maribu** [comment]
This one seems to cause trouble with the rust integration. Looks relatively trivial to fix, though.

**Comment by chrysn** [comment]
I'm just working through the fixing steps, expect a PR updating the Cargo.lock files shortly.

**Pull request merged**